### PR TITLE
Replace calls to methods removed in Django v4

### LINKED ISSUE #1274: Django v4.0 support
Hello!

Django 4.0 is due out next week and I was attempting to test the release candidate with my project. Have there been any efforts to check/fix compatibility with the new version?

### PR DESCRIPTION
**Pull request by MisterGlass**
If I'm running the tests correctly this seems to be all that's needed to resolve #1274

### CODE REVIEW

File: graphene_django/utils/utils.py [hunk 1]
```diff
@@ -26,7 +26,7 @@ def isiterable(value):
 
 def _camelize_django_str(s):
     if isinstance(s, Promise):
-        s = force_text(s)
+        s = force_str(s)
```
Comment (ulgens) [inline graphene_django/utils/utils.py:29]: Is it okay with Django 2.2?

Comment (MisterGlass) [reply graphene_django/utils/utils.py:29]: I'm not sure how I would test that version specifically, but all the project tests pass in my fork https://github.com/MisterGlass/graphene-django/pull/2

Comment (MisterGlass) [reply graphene_django/utils/utils.py:29]: Ah, wait, here is the 2.2 test pass https://github.com/MisterGlass/graphene-django/runs/4385256293?check_suite_focus=true

**Comment by MisterGlass** [comment]
I wasn't running the tests correctly. I ran them manually and found a few more easy changes.

I could use help adding it to the tox config & github actions properly.

File: graphene_django/tests/urls.py [hunk 2]
```diff
@@ -1,8 +1,8 @@
-from django.conf.urls import url
+from django.urls import re_path
 
 from ..views import GraphQLView
 
 urlpatterns = [
-    url(r"^graphql/batch", GraphQLView.as_view(batch=True)),
-    url(r"^graphql", GraphQLView.as_view(graphiql=True)),
+    re_path(r"^graphql/batch", GraphQLView.as_view(batch=True)),
+    re_path(r"^graphql", GraphQLView.as_view(graphiql=True)),
```
Comment (ulgens) [inline graphene_django/tests/urls.py:7]: I think these can be just `path`.

Comment (MisterGlass) [reply graphene_django/tests/urls.py]: I'm not sure why but when I use path here it breaks test_pytest_fixture_usage

**Comment by MisterGlass** [comment]
I updated the tox config & github actions, you can see the various versions tested here: https://github.com/MisterGlass/graphene-django/actions/runs/1533043305

**Comment by tim-schilling** [comment]
@MisterGlass Shouldn't this be against main rather than the v2 branch? The latest version of graphene-django is 3.X

**Comment by MisterGlass** [comment]
> @MisterGlass Shouldn't this be against main rather than the v2 branch? The latest version of graphene-django is 3.X

The latest release on [pypi](https://pypi.org/project/graphene-django/) is 2.15.0. When i started this PR I noticed the 3.0 beta hadn't been updated in a year, so I focused on a 2.x revision.

I'm not sure what state the 3.0 work is in (I'm a new contributor here) but I'd be happy to help make similar changes to it.

**Comment by tim-schilling** [comment]
Ah, I did not see that. You can ignore my comment then.

**Comment by MisterGlass** [comment]
> Ah, I did not see that. You can ignore my comment then.

It probably

**Comment by ulgens** [comment]
> @MisterGlass Shouldn't this be against main rather than the v2 branch? The latest version of graphene-django is 3.X

I agree with this. With the previous Django - Python version update PR, the majority agreed it's too big of an update and it should be part of 3.0. Considering how long it took to release 3.0, i don't think it was the best decision but it's done. At this point, the 2.x version is legacy and I wouldn't expect any new version on that.

About the 3.0 release, I pinged people on Slack a couple of days ago but no one responded yet. Still, I think it's better to contribute to the main branch.

**Comment by MisterGlass** [comment]
Django v4 is now officially released. This library is now the only thing preventing me (and I'm sure others) from upgrading.

The changes I made are minimal, aside from the test configuration I changed 11 lines, replacing deprecated functions with their new equivalents. I really can't see how 11 lines is too big of an update.

There is no mention in the documentation that I can find saying 2.x is legacy & there will be no updates. Was I supposed to be running the beta this whole time? If 2.x is legacy and 3.x is beta, then there is no current maintained version of graphene-django and I need to fork or migrate ASAP.

**Comment by ulgens** [comment]
> If 2.x is legacy and 3.x is beta, then there is no current maintained version of graphene-django and I need to fork or migrate ASAP.

I don't want to discourage anyone from contributing but i'd lying if i say this is not the case.

**Comment by tim-schilling** [comment]
@ulgens, what's blocking us from doing a dual release of 2.x and 3.x?

**Comment by ulgens** [comment]
> @ulgens, what's blocking us from doing a dual release of 2.x and 3.x?

I'd say nothing, but also I'm not in a place to call it.

**Review by zbyte64 — CHANGES_REQUESTED** [review]
We have not abandon our commitment to keep shipping updates to v3, but we also have a commitment to keep up with Django releases. This PR should be accepted once it introduces a version bump (ie 2.16.0) to reflect the drop in support for Django v1. See `graphene_django/__init__.py` for the source of the version number.

We will commit to porting this PR to v3, but I won't be able to get to it right now. Please do us a favor and submit another PR for the main branch, even if it is broken.

**Comment by tim-schilling** [comment]
> We will commit to porting this PR to v3, but I won't be able to get to it right now. Please do us a favor and submit another PR for the main branch, even if it is broken.

This should be done with #1281, though I do need to investigate something regarding `force_text`.

**Comment by cacarrara** [comment]
@zbyte64 sorry for the inconvenient ping. This is just a friendly reminder about this PR. Since your request for changes, MisterGlass updated the version. It would be great to have your approval here.

**Comment by jyotendra** [comment]
@zbyte64 can you plz review the changes and approve the merge. Seems like it's blocking [issue#1284](https://github.com/graphql-python/graphene-django/issues/1284) since long.

**Comment by ulgens** [comment]
Hey everyone, thanks for keeping this alive. I have the permission to merge it if we think it's ready to be merged, but I'd like to have the test matrix cleaned first (removing old versions and simplifying it) If anyone is up for that, please go on. Otherwise, I'll try to handle it as soon as I found some spare time.

**Comment by MisterGlass** [comment]
I'm happy to remove/simplify whatever you think is appropriate for this release, I had originally tried to keep as much compatibility testing as possible.

**Comment by jeremystretch** [comment]
Hi @ulgens! Is there a specific subset of the matrix you had in mind? Here's what it looks like currently in the PR:

| Django      | 2.2 | 3.0 | 3.1 | 3.2 | 4.0 |
|-------------|-----|-----|-----|-----|-----|
| Python 3.6  | x   | x   | x   | x   |     |
| Python 3.7  | x   | x   | x   | x   |     |
| Python 3.8  | x   | x   | x   | x   | x   |
| Python 3.9  | x   | x   | x   | x   | x   |
| Python 3.10 |     |     |     | x   | x   |

IMO there's a string argument for dripping Python 3.6 support as it's been officially deprecated. It might also be reasonable to omit Django 2.2 if we need to further simplify, although that is technically still supported until April.

File: tox.ini [hunk 3]
```diff
@@ -1,24 +1,27 @@
...
 [gh-actions:env]
 DJANGO =
-    1.11: django111
-    2.0: django20
-    2.1: django21
     2.2: django22
     3.0: django30
     3.1: django31
+    3.2: django32
+    3.2: django32
```
Comment (jeremystretch) [inline tox.ini:23]: Looks like a duplicate line.
```suggestion
```

**Comment by ulgens** [comment]
I'd recommend removing 3.6 but keeping 2.2 for now.

File: tox.ini [hunk 4]
```diff
@@ -29,13 +32,11 @@ setenv =
...
-    django111: Django>=1.11,<2.0
-    django111: djangorestframework<3.12
-    django20: Django>=2.0,<2.1
-    django21: Django>=2.1,<2.2
     django22: Django>=2.2,<3.0
-    django30: Django>=3.0a1,<3.1
+    django30: Django>=3.0,<3.1
     django31: Django>=3.1,<3.2
+    django32: Django>=3.2,<4.0
+    django40: Django>=4.0,<5.0
```
Comment (jeremystretch) [inline tox.ini:39]: We probably want to exclude the future 4.1 and 4.2 releases here.
```suggestion
    django40: Django>=4.0,<4.1
```

File: tox.ini [hunk 5]
```diff
@@ -1,24 +1,27 @@
 [tox]
 envlist =
-    py{27,35,36,37,38}-django{111,20,21,22,master},
-    py{36,37,38}-django{30,31},
+    py{35,36,37,38,39}-django22,
```
Comment (jeremystretch) [inline tox.ini:3]: I don't _think_ we need Python 3.5 here.
```suggestion
    py{36,37,38,39}-django22,
```

File: graphene_django/__init__.py [hunk 6]
```diff
@@ -1,7 +1,7 @@
 from .fields import DjangoConnectionField, DjangoListField
 from .types import DjangoObjectType
 
-__version__ = "2.15.0"
+__version__ = "2.16.0"
```
Comment (ulgens) [inline graphene_django/__init__.py:4]: It's better to update when we decided to release.
```suggestion
__version__ = "2.15.0"
```

**Review by ulgens — APPROVED** [review]

**Comment by nextmat** [comment]
Thanks for getting this merged! Just to clarify, what is the likely release plan for this?

**Comment by ulgens** [comment]
> Thanks for getting this merged! Just to clarify, what is the likely release plan for this?

I can't really there is one, but I plan to do some more cleaning under https://github.com/graphql-python/graphene-django/pull/1298 first.

**Pull request merged**